File: app/src/main/java/com/qimian233/ztool/SettingsFragment.java
package com.qimian233.ztool;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.text.Html;
import android.util.Log;
import android.view.HapticFeedbackConstants;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.LinearInterpolator;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.cardview.widget.CardView;
import androidx.core.widget.NestedScrollView;
import androidx.fragment.app.Fragment;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.materialswitch.MaterialSwitch;
import com.qimian233.ztool.hook.modules.SharedPreferencesTool.ModulePreferencesUtils;
import com.qimian233.ztool.service.LogServiceManager;
import com.qimian233.ztool.utils.FileManager;

public class SettingsFragment extends Fragment {

    private ActivityResultLauncher<String> backupLauncher;

    @RequiresApi(api = Build.VERSION_CODES.TIRAMISU)
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_settings, container, false);

        // 获取视图组件
        LinearLayout backupConfigToFile = view.findViewById(R.id.backup_config_to_file);
        LinearLayout restoreConfigFromFile = view.findViewById(R.id.restore_config_from_file);
        LinearLayout restoreDefaultConfig = view.findViewById(R.id.restore_default_config);
        MaterialSwitch switchEnableLogService = view.findViewById(R.id.switch_enable_log_service);
        MaterialSwitch switchEnableDetailedLogging = view.findViewById(R.id.switch_enable_detailed_logging);
        MaterialSwitch switchEnableHomepageYiyan = view.findViewById(R.id.switch_enable_homepage_yiyan);
        CardView showAboutPage = view.findViewById(R.id.show_about_page);

        // 设置点击监听器
        backupConfigToFile.setOnClickListener(v -> performBackup());
        restoreConfigFromFile.setOnClickListener(v ->
                openDocumentLauncherForRestore.launch(new String[]{"application/json"}));
        restoreDefaultConfig.setOnClickListener(v -> restoreDefaultSettings());
        showAboutPage.setOnClickListener(v -> showAboutPage());

        // 设置开关监听器
        boolean isLogServiceEnabled = LogServiceManager.isServiceEnabled(requireContext());
        switchEnableLogService.setChecked(isLogServiceEnabled);
        switchEnableLogService.setOnCheckedChangeListener((buttonView, isChecked) -> handleLogServiceSwitch(isChecked));

        // 这里初始化一个SharedPreferencesUtils实例，需要使用SharedPreferences来初始化配置的开关放在它后面
        ModulePreferencesUtils utils = new ModulePreferencesUtils(requireContext());

        switchEnableDetailedLogging.setChecked(
                utils.loadBooleanSetting("isDetailedLogging", false));
        switchEnableDetailedLogging.setOnCheckedChangeListener(((buttonView, isChecked) ->
                utils.saveBooleanSetting("isDetailedLogging", isChecked)));

        switchEnableHomepageYiyan.setChecked(
                utils.loadBooleanSetting("enable_homepage_yiyan", true));
        switchEnableHomepageYiyan.setOnCheckedChangeListener((buttonView, isChecked) ->
                utils.saveBooleanSetting("enable_homepage_yiyan", isChecked));

        return view;
    }

    private void showToast(String message) {
        if (getContext() != null) {
            android.widget.Toast.makeText(getContext(), message, android.widget.Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        backupLauncher = registerForActivityResult(
                new ActivityResultContracts.CreateDocument("application/json"), uri -> {
                    if (uri != null) {
                        boolean result = FileManager.saveConfigWithSAF(requireContext(),
                                uri,
                                FileManager.generateBackupFileName(),
                                ModulePreferencesUtils.getAllSettingsAsJSON(requireContext()));
                        if (result) {
                            Log.d("SAF", "成功存储配置到用户指定的目录" + uri);
                            showToast(getString(R.string.config_backup_success));
                        } else {
                            Log.e("SAF", "备份失败");
                        }
                    }
                });
    }

    @Override
    public void onResume() {
        super.onResume();
        refreshSwitchStates();
    }

    private void refreshSwitchStates() {
        if (getContext() == null || getView() == null) return;
        boolean isLogServiceEnabled = LogServiceManager.isServiceEnabled(requireContext());
        MaterialSwitch switchEnableLogService = requireView().findViewById(R.id.switch_enable_log_service);
        if (switchEnableLogService != null) {
            switchEnableLogService.setChecked(isLogServiceEnabled);
        }
    }

    // =======================================================
    // 重写的关于页面逻辑 (全屏沉浸式 + 点击任意空白关闭)
    // =======================================================

    private void showAboutPage() {
        // 1. 初始化 View
        View dialogView = LayoutInflater.from(requireContext()).inflate(R.layout.dialog_about_custom, null);

        View backgroundView = dialogView.findViewById(R.id.about_background);
        NestedScrollView scrollView = dialogView.findViewById(R.id.about_scroll_view);
        LinearLayout contentContainer = dialogView.findViewById(R.id.about_content_container);

        View iconCard = dialogView.findViewById(R.id.about_icon_card);
        TextView titleView = dialogView.findViewById(R.id.about_title);
        TextView versionView = dialogView.findViewById(R.id.about_version);
        TextView descView = dialogView.findViewById(R.id.about_description);
        View btnContainerProject = dialogView.findViewById(R.id.btn_container_project_info);
        View btnContainerAuthor = dialogView.findViewById(R.id.btn_container_author_info);
        MaterialButton btnGithub = dialogView.findViewById(R.id.btn_github);
        MaterialButton btnCredits = dialogView.findViewById(R.id.btn_credits);
        MaterialButton btnAuthor = dialogView.findViewById(R.id.btn_qimian233_coolapk);
        MaterialButton btnCollaborator = dialogView.findViewById(R.id.btn_wasddestroy_coolapk);

        // 2. 设置数据
        versionView.setText(updateModuleStatus());
        String rawHtml = getString(R.string.about_description);
        String plainText = Html.fromHtml(rawHtml, Html.FROM_HTML_MODE_LEGACY).toString();

        btnGithub.setOnClickListener(v ->
                openExternalLink("https://github.com/qwqawa64/ZUX-ZTool",
                        false, ""));

        btnCredits.setOnClickListener(v ->
                openExternalLink("https://github.com/dantmnf/UnfuckZUI",
                false, ""));

        btnAuthor.setOnClickListener(v -> openExternalLink("http://www.coolapk.com/u/10099756",
                true, "com.coolapk.market")); // Qimian233

        btnCollaborator.setOnClickListener(v -> openExternalLink("http://www.coolapk.com/u/18634835",
                true, "com.coolapk.market")); // WASD_Destroy

        // 3. 创建 Dialog
        AlertDialog dialog = new MaterialAlertDialogBuilder(requireContext())
                .setView(dialogView)
                .setCancelable(true)
                .create();

        // 4. 【关键修改】设置点击关闭逻辑
        // 定义一个通用的关闭监听器
        View.OnClickListener dismissListener = v -> dialog.dismiss();

        // 给底层背景设置关闭
        backgroundView.setOnClickListener(dismissListener);
        // 给滚动视图层设置关闭 (因为 match_parent 会遮挡背景)
        scrollView.setOnClickListener(dismissListener);
        contentContainer.setOnClickListener(dismissListener);

        // 5. 显示 Dialog 并配置 Window 参数
        dialog.show();
        Window window = dialog.getWindow();
        if (window != null) {
            window.setLayout(WindowManager.LayoutParams.MATCH_PARENT, WindowManager.LayoutParams.MATCH_PARENT);
            window.setDimAmount(0f);
            window.setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        }

        // ============ 动画编排 ============

        long startDelay = 50;
        long charInterval = 25;
        long totalTypingTime = plainText.length() * charInterval;

        // 背景渐变动画
        backgroundView.setAlpha(0f);
        backgroundView.animate()
                .alpha(1.0f)
                .setDuration(300)
                .setStartDelay(0)
                .setInterpolator(new LinearInterpolator())
                .start();

        // 内容元素入场
        animateEntrance(iconCard, startDelay);
        animateEntrance(titleView, startDelay + 100);
        animateEntrance(versionView, startDelay + 200);

        // 打字机文字
        new Handler(Looper.getMainLooper()).postDelayed(() ->
                        typeWriterAnimation(descView, plainText, charInterval),
                startDelay + 300
        );

        // 底部按钮最后浮现
        animateEntrance(btnContainerProject, startDelay + totalTypingTime + 200);
        animateEntrance(btnContainerAuthor, startDelay + totalTypingTime + 400);
    }

    private void animateEntrance(View view, long delay) {
        if (view == null) return;
        view.setAlpha(0f);
        view.setTranslationY(60f);

        view.animate()
                .alpha(1f)
                .translationY(0f)
                .setStartDelay(delay)
                .setDuration(600)
                .setInterpolator(new DecelerateInterpolator(1.5f))
                .start();
    }

    private void typeWriterAnimation(TextView textView, String text, long interval) {
        if (textView == null || text == null) return;
        final Handler handler = new Handler(Looper.getMainLooper());
        final int length = text.length();
        final StringBuilder builder = new StringBuilder();

        for (int i = 0; i < length; i++) {
            final int index = i;
            handler.postDelayed(() -> {
                if (!isAdded() || getContext() == null) return;
                builder.append(text.charAt(index));
                textView.setText(builder.toString());

                if (index % 2 == 0) {
                    textView.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP);
                }
            }, i * interval);
        }
    }

    // =======================================================

    private void restoreDefaultSettings() {
        AlertDialog dialog = new MaterialAlertDialogBuilder(requireContext())
                .setTitle(R.string.final_confirmation_title)
                .setMessage(R.string.restore_default_confirmation)
                .setPositiveButton(R.string.restart_yes,
                        (dialogInterface, which) -> performRestore())
                .setNegativeButton(R.string.restart_no, null)
                .create();
        dialog.show();
    }

    private void performRestore() {
        new ModulePreferencesUtils(requireContext()).clearAllSettings();
        showToast(getString(R.string.default_config_restored));
    }

    private void performBackup() {
        String backupFileName = FileManager.generateBackupFileName();
        backupLauncher.launch(backupFileName);
    }

    private void openExternalLink(String link, boolean shouldDeterminePackage, String packageName) {
        try {
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(link));
            if (shouldDeterminePackage) {
                intent.setPackage(packageName);
            }
            startActivity(intent);
        } catch (Exception e) {
            showToast(getString(R.string.open_web_link_failed));
        }
    }

    private final ActivityResultLauncher<String[]> openDocumentLauncherForRestore =
            registerForActivityResult(new ActivityResultContracts.OpenDocument(), uri -> {
                if (uri != null) {
                    String content = FileManager.readConfigWithSAF(requireContext(), uri);
                    if (content != null) {
                        Log.d("SAF", "读取到的内容: " + content);
                        ModulePreferencesUtils.restoreConfig(requireContext(), content);
                        showToast(getString(R.string.config_restore_success));
                    } else {
                        Log.e("SAF", "文件读取失败或内容为空");
                    }
                }
            });

    private void handleLogServiceSwitch(boolean isEnabled) {
        if (isEnabled) {
            LogServiceManager.startLogService(requireContext());
            showToast(getString(R.string.log_service_started));
        } else {
            LogServiceManager.stopLogService(requireContext());
            showToast(getString(R.string.log_service_stopped));
        }
    }

    private String updateModuleStatus() {
        final String TAG = "SettingsFragment";
        String moduleVersion;
        try {
            try {
                Activity activity = getActivity();
                if (activity != null) {
                    PackageInfo packageInfo = activity.getPackageManager().getPackageInfo(activity.getPackageName(), 0);
                    String versionName = packageInfo.versionName;
                    int versionCode = packageInfo.versionCode;
                    moduleVersion = versionName + " (" + versionCode + ")";
                } else {
                    moduleVersion = getString(R.string.unknown_activity_null);
                }
            } catch (PackageManager.NameNotFoundException e) {
                Log.w(TAG, "无法获取模块版本信息: " + e.getMessage());
                moduleVersion = getString(R.string.unknown);
            }
            Log.i(TAG, "模块状态更新完成");
            return moduleVersion;

        } catch (Exception e) {
            Log.e(TAG, "更新模块状态失败: " + e.getMessage());
            return null;
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/SettingsFragment.kt
package com.qimian233.ztool

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.rounded.ArrowForward
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.qimian233.ztool.data.settings.SettingsRepository
import com.qimian233.ztool.ui.components.ZToolCard
import com.qimian233.ztool.ui.components.ZToolDialog
import com.qimian233.ztool.ui.components.ZToolPageSurface
import com.qimian233.ztool.ui.components.ZToolSwitchRow
import com.qimian233.ztool.ui.theme.ZToolTheme
import com.qimian233.ztool.viewmodel.SettingsUiState
import com.qimian233.ztool.viewmodel.SettingsViewModel

class SettingsFragment : Fragment() {

    private lateinit var viewModel: SettingsViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val repository = SettingsRepository(requireContext().applicationContext)
        viewModel = ViewModelProvider(
            this,
            SettingsViewModelFactory(repository)
        )[SettingsViewModel::class.java]
    }

    private val backupLauncher = registerForActivityResult(
        ActivityResultContracts.CreateDocument("application/json")
    ) { uri ->
        if (uri != null) {
            viewModel.backupConfig(uri) { result ->
                activity?.runOnUiThread {
                    if (result) {
                        showToast(getString(R.string.config_backup_success))
                    }
                }
            }
        }
    }

    private val restoreLauncher = registerForActivityResult(
        ActivityResultContracts.OpenDocument()
    ) { uri ->
        if (uri != null) {
            viewModel.restoreConfig(uri) { result ->
                activity?.runOnUiThread {
                    if (result) {
                        showToast(getString(R.string.config_restore_success))
                    }
                }
            }
        }
    }

    override fun onCreateView(
        inflater: android.view.LayoutInflater,
        container: android.view.ViewGroup?,
        savedInstanceState: Bundle?
    ): android.view.View {
        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                val uiState by viewModel.uiState.collectAsState()

                ZToolTheme {
                    SettingsRoute(
                        state = uiState,
                        onBackup = { backupLauncher.launch(viewModel.backupFileName()) },
                        onRestore = { restoreLauncher.launch(arrayOf("application/json")) },
                        onRestoreDefault = viewModel::showRestoreConfirmDialog,
                        onLogServiceChanged = {
                            viewModel.setLogServiceEnabled(it)
                            showToast(
                                getString(
                                    if (it) R.string.log_service_started else R.string.log_service_stopped
                                )
                            )
                        },
                        onDetailedLoggingChanged = viewModel::setDetailedLoggingEnabled,
                        onHomepageYiyanChanged = viewModel::setHomepageYiyanEnabled,
                        onAbout = viewModel::showAboutDialog
                    )

                    if (uiState.showRestoreConfirmDialog) {
                        RestoreDefaultDialog(
                            onConfirm = {
                                viewModel.restoreDefaultConfig()
                                showToast(getString(R.string.default_config_restored))
                            },
                            onDismiss = viewModel::dismissRestoreConfirmDialog
                        )
                    }

                    if (uiState.showAboutDialog) {
                        AboutDialog(
                            version = uiState.moduleVersion,
                            onDismiss = viewModel::dismissAboutDialog,
                            onOpenGithub = {
                                openExternalLink("https://github.com/qwqawa64/ZUX-ZTool", false, "")
                            },
                            onOpenCredits = {
                                openExternalLink("https://github.com/dantmnf/UnfuckZUI", false, "")
                            },
                            onOpenAuthor = {
                                openExternalLink("http://www.coolapk.com/u/10099756", true, "com.coolapk.market")
                            },
                            onOpenCollaborator = {
                                openExternalLink("http://www.coolapk.com/u/18634835", true, "com.coolapk.market")
                            }
                        )
                    }
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        viewModel.refresh()
    }

    private fun openExternalLink(link: String, shouldDeterminePackage: Boolean, packageName: String) {
        try {
            startActivity(
                Intent(Intent.ACTION_VIEW, Uri.parse(link)).apply {
                    if (shouldDeterminePackage) setPackage(packageName)
                }
            )
        } catch (_: Exception) {
            showToast(getString(R.string.open_web_link_failed))
        }
    }

    private fun showToast(message: String) {
        if (context != null) {
            Toast.makeText(requireContext(), message, Toast.LENGTH_SHORT).show()
        }
    }

}

private class SettingsViewModelFactory(
    private val repository: SettingsRepository
) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(SettingsViewModel::class.java)) {
            return SettingsViewModel(repository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class: ${modelClass.name}")
    }
}

@Composable
private fun SettingsRoute(
    state: SettingsUiState,
    onBackup: () -> Unit,
    onRestore: () -> Unit,
    onRestoreDefault: () -> Unit,
    onLogServiceChanged: (Boolean) -> Unit,
    onDetailedLoggingChanged: (Boolean) -> Unit,
    onHomepageYiyanChanged: (Boolean) -> Unit,
    onAbout: () -> Unit
) {
    ZToolPageSurface(
        modifier = Modifier.fillMaxSize(),
        contentAlignment = Alignment.TopCenter
    ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .widthIn(max = 960.dp)
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 32.dp, vertical = 32.dp)
        ) {
            Text(
                text = stringResource(R.string.settingsFragment_title),
                style = MaterialTheme.typography.headlineMedium,
                color = MaterialTheme.colorScheme.onSurface
            )
            Spacer(modifier = Modifier.height(24.dp))

            SettingsSection(title = stringResource(R.string.backupAndRestore)) {
                SettingsActionRow(
                    title = stringResource(R.string.backupConfigToFile),
                    onClick = onBackup
                )
                SettingsActionRow(
                    title = stringResource(R.string.restoreConfigFromFile),
                    onClick = onRestore
                )
                SettingsActionRow(
                    title = stringResource(R.string.restoreDefaultConfig),
                    onClick = onRestoreDefault
                )
            }

            Spacer(modifier = Modifier.height(16.dp))

            SettingsSection(title = stringResource(R.string.moreSettings)) {
                ZToolSwitchRow(
                    title = stringResource(R.string.enableLogService),
                    summary = stringResource(R.string.enableLogServiceDescription),
                    checked = state.isLogServiceEnabled,
                    onCheckedChange = onLogServiceChanged
                )
                ZToolSwitchRow(
                    title = stringResource(R.string.enableDetailedLogging),
                    summary = stringResource(R.string.enableDetailedLoggingDescription),
                    checked = state.isDetailedLoggingEnabled,
                    onCheckedChange = onDetailedLoggingChanged
                )
                ZToolSwitchRow(
                    title = stringResource(R.string.enableHomePageYiyan),
                    summary = stringResource(R.string.enableHomePageYiyanSummary),
                    checked = state.isHomepageYiyanEnabled,
                    onCheckedChange = onHomepageYiyanChanged
                )
            }

            Spacer(modifier = Modifier.height(16.dp))

            SettingsSection {
                SettingsActionRow(
                    title = stringResource(R.string.showAboutPage),
                    onClick = onAbout
                )
            }
            Spacer(modifier = Modifier.height(32.dp))
        }
    }
}

@Composable
private fun SettingsSection(
    title: String? = null,
    content: @Composable ColumnScope.() -> Unit
) {
    ZToolCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 12.dp)
        ) {
            if (title != null) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleLarge,
                    color = MaterialTheme.colorScheme.onSurface,
                    modifier = Modifier.padding(horizontal = 20.dp, vertical = 8.dp)
                )
            }
            content()
        }
    }
}

@Composable
private fun SettingsActionRow(
    title: String,
    onClick: () -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(horizontal = 20.dp, vertical = 16.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = title,
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.onSurface,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f)
        )
        Spacer(modifier = Modifier.width(16.dp))
        Icon(
            imageVector = Icons.AutoMirrored.Rounded.ArrowForward,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

@Composable
private fun RestoreDefaultDialog(
    onConfirm: () -> Unit,
    onDismiss: () -> Unit
) {
    ZToolDialog(
        onDismissRequest = onDismiss,
        title = { Text(stringResource(R.string.final_confirmation_title)) },
        text = { Text(stringResource(R.string.restore_default_confirmation)) },
        confirmButton = {
            TextButton(onClick = onConfirm) {
                Text(stringResource(R.string.restart_yes))
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text(stringResource(R.string.restart_no))
            }
        }
    )
}

@Composable
private fun AboutDialog(
    version: String,
    onDismiss: () -> Unit,
    onOpenGithub: () -> Unit,
    onOpenCredits: () -> Unit,
    onOpenAuthor: () -> Unit,
    onOpenCollaborator: () -> Unit
) {
    ZToolDialog(
        onDismissRequest = onDismiss,
        title = { Text(stringResource(R.string.about_ztool_title)) },
        text = {
            Column {
                Text(
                    text = version,
                    style = MaterialTheme.typography.titleSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(modifier = Modifier.height(12.dp))
                Text(
                    text = stringResource(R.string.about_description)
                        .replace("<br>", "\n")
                        .replace("&lt;br&gt;", "\n"),
                    style = MaterialTheme.typography.bodyMedium
                )
            }
        },
        confirmButton = {
            TextButton(onClick = onDismiss) {
                Text(stringResource(android.R.string.ok))
            }
        },
        dismissButton = {
            Row {
                TextButton(
                    contentPadding = PaddingValues(horizontal = 8.dp),
                    onClick = onOpenGithub
                ) {
                    Text(stringResource(R.string.button_project_homepage))
                }
                TextButton(
                    contentPadding = PaddingValues(horizontal = 8.dp),
                    onClick = onOpenCredits
                ) {
                    Text("Credits")
                }
                TextButton(
                    contentPadding = PaddingValues(horizontal = 8.dp),
                    onClick = onOpenAuthor
                ) {
                    Text("Qimian233")
                }
                TextButton(
                    contentPadding = PaddingValues(horizontal = 8.dp),
                    onClick = onOpenCollaborator
                ) {
                    Text("WASD")
                }
            }
        }
    )
}
